src/logion/controllers/secret_recovery.controller.ts

```diff
@@ -0,0 +1,138 @@
+    Log
+} from "@logion/rest-api-core";
+import { injectable } from "inversify";
+import { Controller, HttpPost, ApiController, Async, SendsResponse } from "dinoloop";
+import { components } from "./components.js";
+import { SecretRecoveryRequestFactory, SecretRecoveryRequestDescription } from "../model/secret_recovery.model.js";
+import { SecretRecoveryRequestService } from "../services/secret_recovery.service.js";
+import { LocRequestRepository } from "../model/locrequest.model.js";
+import moment from "moment";
+import { NotificationRecipient, Template, NotificationService } from "../services/notification.service.js";
+import { UserPrivateData } from "./adapters/locrequestadapter.js";
+import { LocalsObject } from "pug";
+import { DirectoryService } from "../services/directory.service.js";
+import { ValidAccountId } from "@logion/node-api";
+
+type CreateSecretRecoveryRequestView = components["schemas"]["CreateSecretRecoveryRequestView"];
+
+const { logger } = Log;
+
+export function fillInSpec(spec: OpenAPIV3.Document): void {
+    const tagName = 'SecretRecovery';
+    addTag(spec, {
+        name: tagName,
+        description: "Handling of Secret Recovery Requests"
+    });
+    setControllerTag(spec, /^\/api\/secret-recovery.*/, tagName);
+
+    SecretRecoveryController.createSecretRecoveryRequest(spec);
+}
+
+@injectable()
+@Controller('/secret-recovery')
+export class SecretRecoveryController extends ApiController {
+
+    constructor(
+        private secretRecoveryRequestFactory: SecretRecoveryRequestFactory,
+        private secretRecoveryRequestService: SecretRecoveryRequestService,
+        private locRequestRepository: LocRequestRepository,
+        private directoryService: DirectoryService,
+        private notificationService: NotificationService,
+    ) {
+        super();
+    }
+
+    static createSecretRecoveryRequest(spec: OpenAPIV3.Document) {
+        const operationObject = spec.paths["/api/secret-recovery"].post!;
+        operationObject.summary = "Creates a new Secret recovery request";
+        operationObject.description = "This is publicly available";
+        operationObject.requestBody = getRequestBody({
+            description: "Secret recovery request creation data",
+            view: "secret-recovery",
+        });
+        operationObject.responses = getDefaultResponses("ProtectionRequestView");
+    }
+
+    @Async()
+    @HttpPost('')
+    @SendsResponse()
+    async createSecretRecoveryRequest(body: CreateSecretRecoveryRequestView) {
+        const requesterIdentityLoc = requireDefined(
+            await this.locRequestRepository.findById(body.requesterIdentityLocId),
+            () => badRequest("Identity LOC not found")
+        );
+        requireDefined(
+            requesterIdentityLoc.secrets?.find(secret => secret.name === body.secretName),
+            () => badRequest("Secret not found")
+        )
+        const { requesterIdentityLocId, challenge, secretName } = body;
```

Comment:
This statement should be moved at the beginning of the block so that variables can be used for the above queries.

Reply:
Sure.